collector/lib/CollectorConfig.cpp

```diff
@@ -422,33 +425,106 @@ void CollectorConfig::YamlConfigToConfig(YAML::Node& yamlConfig) {
   if (!std::filesystem::exists(filePath)) {
     CLOG(DEBUG) << "No configuration file found. " << filePath;
-    return;
+    return true;
   }
 
   try {
     YAML::Node yamlConfig = YAML::LoadFile(filePath);
     YamlConfigToConfig(yamlConfig);
   } catch (const YAML::BadFile& e) {
-    CLOG(FATAL) << "Failed to open the configuration file: " << filePath
-                << ". Error: " << e.what();
+    CLOG(ERROR) << "Failed to open the configuration file: " << filePath << ". Error: " << e.what();
+    return false;
   } catch (const YAML::ParserException& e) {
-    CLOG(FATAL) << "Failed to parse the configuration file: " << filePath
-                << ". Error: " << e.what();
+    CLOG(ERROR) << "Failed to parse the configuration file: " << filePath << ". Error: " << e.what();
+    return false;
   } catch (const YAML::Exception& e) {
-    CLOG(FATAL) << "An error occurred while loading the configuration file: " << filePath
-                << ". Error: " << e.what();
+    CLOG(ERROR) << "An error occurred while loading the configuration file: " << filePath << ". Error: " << e.what();
+    return false;
   } catch (const std::exception& e) {
-    CLOG(FATAL) << "An unexpected error occurred while trying to read: " << filePath << e.what();
+    CLOG(ERROR) << "An unexpected error occurred while trying to read: " << filePath << e.what();
+    return false;
+  }
+
+  return true;
+}
+
+void CollectorConfig::WaitForFileToExist(const std::filesystem::path& filePath) {
+  while (!std::filesystem::exists(filePath)) {
+    if (runtime_config_.has_value()) {
+      CLOG(INFO) << "The configuration file has been deleted. Resetting the configuration.";
+      runtime_config_.reset();
+    }
+    sleep(1);
+  }
+}
+
+int WaitForInotifyAddWatch(int fd, const std::filesystem::path& filePath) {
+  while (true) {
+    int wd = inotify_add_watch(fd, filePath.c_str(), IN_MODIFY | IN_MOVE_SELF | IN_DELETE_SELF);
+    if (wd < 0) {
+      CLOG_THROTTLED(ERROR, std::chrono::seconds(30)) << "Failed to add inotify watch for " << filePath;
+      close(fd);
+    } else {
+      return wd;
+    }
+    sleep(1);
+  }
+}
+
+void CollectorConfig::WatchConfigFile(const std::filesystem::path& filePath) {
+  int fd = inotify_init();
+  if (fd < 0) {
+    CLOG(ERROR) << "inotify_init() failed: " << StrError();
+    CLOG(ERROR) << "Runtime configuration will not be used.";
+    return;
+  }
+  WaitForFileToExist(filePath);
+  int wd = WaitForInotifyAddWatch(fd, filePath);
+  bool success = HandleConfig(filePath);
+  if (!success) {
+    CLOG(FATAL) << "Unable to parse configuration file: " << filePath;
   }
+
+  char buffer[1024];
+  while (true) {
+    int length = read(fd, buffer, sizeof(buffer));
+    if (length < 0) {
+      CLOG(ERROR) << "Unable to read event for " << filePath;
+    }
+
+    for (int i = 0; i < length; i += sizeof(struct inotify_event)) {
+      struct inotify_event* event = (struct inotify_event*)&buffer[i];
```

Comment:
inotify events have a variable length array in them, you need to adjust for that here
```suggestion
    struct inotify_event *event = buffer;
    for (int i = 0; i < length; i += sizeof(struct inotify_event) + event->len) {
      event = (struct inotify_event*)&buffer[i];
```